pkg/alertmanager/multitenant.go

```diff
@@ -243,65 +246,70 @@ func (am *MultitenantAlertmanager) Stop() {
+func (am *MultitenantAlertmanager) loadAllConfigs() map[string]alerts.AlertConfigDesc {
 	backoff := util.NewBackoff(context.Background(), backoffConfig)
 	for {
 		cfgs, err := am.poll()
 		if err == nil {
-			level.Debug(util.Logger).Log("msg", "MultitenantAlertmanager: initial configuration load", "num_configs", len(cfgs))
+			level.Debug(am.logger).Log("msg", "initial configuration load", "num_configs", len(cfgs))
 			return cfgs
 		}
-		level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: error fetching all configurations, backing off", "err", err)
+		level.Warn(am.logger).Log("msg", "error fetching all configurations, backing off", "err", err)
 		backoff.Wait()
 	}
 }
 
-func (am *MultitenantAlertmanager) updateConfigs(now time.Time) error {
+func (am *MultitenantAlertmanager) updateConfigs() error {
 	cfgs, err := am.poll()
 	if err != nil {
 		return err
 	}
-	am.addNewConfigs(cfgs)
+	am.syncConfigs(cfgs)
 	return nil
 }
 
-// poll the configuration server. Not re-entrant.
-func (am *MultitenantAlertmanager) poll() (map[string]configs.View, error) {
-	configID := am.latestConfig
-	cfgs, err := am.configsAPI.GetAlerts(context.Background(), configID)
+// poll the alert store. Not re-entrant.
+func (am *MultitenantAlertmanager) poll() (map[string]alerts.AlertConfigDesc, error) {
+	cfgs, err := am.store.ListAlertConfigs(context.Background())
 	if err != nil {
-		level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: configs server poll failed", "err", err)
+		level.Warn(am.logger).Log("msg", "configs alert store poll failed", "err", err)
 		return nil, err
 	}
-	am.latestMutex.Lock()
-	am.latestConfig = cfgs.GetLatestConfigID()
-	am.latestMutex.Unlock()
-	return cfgs.Configs, nil
+	return cfgs, nil
 }
 
-func (am *MultitenantAlertmanager) addNewConfigs(cfgs map[string]configs.View) {
-	// TODO: instrument how many configs we have, both valid & invalid.
-	level.Debug(util.Logger).Log("msg", "adding configurations", "num_configs", len(cfgs))
-	for userID, config := range cfgs {
-		if config.IsDeleted() {
-			am.deleteUser(userID)
-			continue
-		}
-		err := am.setConfig(userID, config.Config)
+func (am *MultitenantAlertmanager) syncConfigs(cfgs map[string]alerts.AlertConfigDesc) {
+	invalid := 0 // Count the number of invalid configs as we go.
+
+	level.Debug(am.logger).Log("msg", "adding configurations", "num_configs", len(cfgs))
+	for _, cfg := range cfgs {
+		err := am.setConfig(cfg)
 		if err != nil {
-			level.Warn(util.Logger).Log("msg", "MultitenantAlertmanager: error applying config", "err", err)
+			invalid++
+			level.Warn(am.logger).Log("msg", "error applying config", "err", err)
 			continue
 		}
+	}
 
+	am.alertmanagersMtx.Lock()
+	defer am.alertmanagersMtx.Unlock()
+	for user, userAM := range am.alertmanagers {
+		if _, exists := cfgs[user]; !exists {
+			userAM.Stop()
+			delete(am.alertmanagers, user)
+			delete(am.cfgs, user)
+			level.Info(am.logger).Log("msg", "deleting alertmanager", "user", user)
```

Comment:
This log should be moved before `userAM.Stop()` because it says "deleting" (here the deletion has been completed). I'm OK being verbose and keeping this log as well, but it should say "deleted alertmanager"

Reply:
I thought I had already updated this line to have two separate log lines. `deleting/deleted`. Guess not, I will fix it to have two separate lines.